### PR TITLE
Automatically retry EasyPost Shipment creation for common errors

### PR DESCRIPTION
**Pull request by neelk07**
TrackerUrl: https://www.pivotaltracker.com/story/show/167745836

### CODE REVIEW

File: app/models/spree/shipment_decorator.rb [hunk 1]
```diff
@@ -35,12 +35,58 @@ def build_easypost_shipment
...
+      easypost_shipment = retrieve_or_build_easypost_shipment
+
       rate = easypost_shipment.rates.find do |rate|
         rate.id == selected_easy_post_rate_id
       end
 
-      easypost_shipment.buy(rate) unless easypost_shipment.postage_label.present?
+      begin
+        easypost_shipment.buy(rate) unless easypost_shipment.postage_label.present?
+      rescue Exception => exception
```
Comment (PeteTheSadPanda) [inline app/models/spree/shipment_decorator.rb]: RuntimeError, never rescue just Exception

File: app/models/spree/shipment_decorator.rb [hunk 2]
```diff
@@ -35,12 +35,58 @@ def build_easypost_shipment
...
+        old_rate = self.shipping_rates.find_by(name: "#{rate.carrier} #{rate.service}")
+        for_selected_rate = "#{rate.carrier} #{rate.service}" == self.selected_shipping_rate.name
+
+        if old_rate.blank?
+          return if for_selected_rate
+          next
+        end
+
+        difference_in_price = rate.rate.to_f - old_rate.cost.to_f
+        for_selected_rate = "#{rate.carrier} #{rate.service}" == self.selected_shipping_rate.name
```
Comment (bengolden) [inline app/models/spree/shipment_decorator.rb]: it looks like this line is a duplicate of line 48 and not needed

File: app/models/spree/shipment_decorator.rb [hunk 3]
```diff
@@ -35,12 +35,58 @@ def build_easypost_shipment
       )
     end
 
+    def rebuild_easypost_shipment
+      new_ep_shipment = build_easypost_shipment
+      new_ep_rates = new_ep_shipment.rates
+      return if new_ep_shipment.messages.present? || new_ep_rates.count < self.shipping_rates.count
+
+      matches = {}
+      no_matches_found = []
```
Comment (bengolden) [inline app/models/spree/shipment_decorator.rb]: this variable is unused

File: app/models/spree/shipment_decorator.rb [hunk 4]
```diff
@@ -35,12 +35,58 @@ def build_easypost_shipment
...
+
+      new_ep_rates.each do |rate|
+        old_rate = self.shipping_rates.find_by(name: "#{rate.carrier} #{rate.service}")
+        for_selected_rate = "#{rate.carrier} #{rate.service}" == self.selected_shipping_rate.name
+
+        if old_rate.blank?
```
Comment (PeteTheSadPanda) [inline app/models/spree/shipment_decorator.rb:49]: this seems like it's just if old_rate.blank? && for_selected_rate

File: app/models/spree/shipment_decorator.rb [hunk 5]
```diff
@@ -35,12 +35,58 @@ def build_easypost_shipment
...
+          return if for_selected_rate
+          next
+        end
+
+        difference_in_price = rate.rate.to_f - old_rate.cost.to_f
+        for_selected_rate = "#{rate.carrier} #{rate.service}" == self.selected_shipping_rate.name
+        return if difference_in_price != 0.0 && for_selected_rate
+
+        matches["#{old_rate.easy_post_rate_id}"] = rate.id
```
Comment (PeteTheSadPanda) [inline app/models/spree/shipment_decorator.rb]: matches[old_rate.easy_post_rate_id.to_s]

**Review by PeteTheSadPanda — CHANGES_REQUESTED** [review]

**Comment by PeteTheSadPanda** [comment]
Looks good. Merging.

**Pull request merged**